src/shogun/lib/any.h

```diff
@@ -0,0 +1,291 @@
+
+        /** Constructor to copy value */
+        template <typename T>
+        explicit Any(const T& v) : policy(selectPolicy<T>()), storage(nullptr)
+        {
+            policy->set(&storage, &v);
+        }
+
+        /** Copy constructor */
+        Any(const Any& other) : policy(other.policy), storage(nullptr)
+        {
+            policy->set(&storage, other.storage);
+        }
+
+        /** Assignment operator */
+        Any& operator=(const Any& other)
+        {
+            policy->clear(&storage);
+            policy = other.policy;
+            policy->set(&storage, other.storage);
+            return *(this);
+        }
+
+        /** Equality operator */
+        friend inline bool operator==(const Any& lhs, const Any& rhs);
+
+        /** Inequality operator */
+        friend inline bool operator!=(const Any& lhs, const Any& rhs);
+
+        /** Destructor */
+        ~Any()
+        {
+            policy->clear(&storage);
+        }
+
+        /** Casts hidden value to provided type, fails otherwise.
+         * @return type-casted value
+         */
+        template <typename T>
+        T& as() const
+        {
+            if (sameType<T>())
+            {
+                return *(reinterpret_cast<T*>(storage));
+            } 
+            else 
+            {
+                throw std::logic_error("Bad cast to " + demangledType<T>() + 
+                        " but the type is " + policy->type());
+            }
+        }
+
+        /** @return true if type is same. */
+        template <typename T>
+        inline bool sameType() const
+        {
+            return (policy == selectPolicy<T>()) || sameTypeFallback<T>();
+        }
+
+        /** @return true if type-id is same. */
+        template <typename T>
+        bool sameTypeFallback() const
+        {
+            return policy->matches(typeid(T));
+        }
+
+        /** @return true if Any object is empty. */
+        bool empty() const
+        {
+            return sameType<Empty>();
+        }
+    private:
+        template <typename T>
+        static BaseAnyPolicy* selectPolicy()
+        {
+            typedef PointerValueAnyPolicy<T> Policy;
+            static Policy policy;
+            return &policy;
+        }
+
+        BaseAnyPolicy* policy;
+        void* storage;
+    };
+
+    inline bool operator==(const Any& lhs, const Any& rhs)
+    {
+        void* lhs_storage = lhs.storage;
+        void* rhs_storage = rhs.storage;        
+        return lhs.policy == rhs.policy and
+            lhs.policy->equals(&lhs_storage, &rhs_storage);
+    }
+
+    inline bool operator!=(const Any& lhs, const Any& rhs)
+    {        
+        return !(lhs == rhs);
+    }
+
+    /** Used to denote an empty Any object */
+    struct Any::Empty
+    {
+        /** Equality operator */
+        bool operator==(const Empty& other) const
+        {
+            return true;
+        }
+    };
+
+    /** Erases value type i.e. converts it to Any
```

Comment:
I don't understand why converting to Any means erasing. Can you explain that in the docs?